test/healthcheck_test.go

```diff
@@ -120,17 +112,69 @@ func (s *testHealthServer) SetServingStatus(service string, status healthpb.Heal
 	s.mu.Unlock()
 }
 
-func TestHealthCheckWatchStateChange(t *testing.T) {
-	defer leakcheck.Check(t)
-	s := grpc.NewServer()
-	lis, err := net.Listen("tcp", "localhost:0")
+func setupHealthCheckWrapper() (hcEnterChan chan struct{}, hcExitChan chan struct{}, wrapper func(ctx context.Context, newStream func() (interface{}, error), update func(bool), service string) error) {
+	hcEnterChan = make(chan struct{})
+	hcExitChan = make(chan struct{})
+	wrapper = func(ctx context.Context, newStream func() (interface{}, error), update func(bool), service string) error {
+		close(hcEnterChan)
+		defer close(hcExitChan)
+		return testHealthCheckFunc(ctx, newStream, update, service)
+	}
+	return
+}
+
+type svrConfig struct {
+	specialWatchFunc func(s *testHealthServer, in *healthpb.HealthCheckRequest, stream healthgrpc.Health_WatchServer) error
+}
+
+func setupServer(sc *svrConfig) (s *grpc.Server, lis net.Listener, ts *testHealthServer, deferFunc func(), err error) {
+	s = grpc.NewServer()
+	lis, err = net.Listen("tcp", "localhost:0")
 	if err != nil {
-		t.Fatalf("failed to listen due to err: %v", err)
+		return nil, nil, nil, func() {}, fmt.Errorf("failed to listen due to err %v", err)
+	}
+	if sc.specialWatchFunc != nil {
+		ts = newTestHealthServerWithWatchFunc(sc.specialWatchFunc)
+	} else {
+		ts = newTestHealthServer()
 	}
-	ts := newTestHealthServer()
 	healthgrpc.RegisterHealthServer(s, ts)
+	testpb.RegisterTestServiceServer(s, &testServer{})
 	go s.Serve(lis)
-	defer s.Stop()
+	return s, lis, ts, func() { s.Stop() }, nil
+}
+
+type clientConfig struct {
+	balancerName               string
+	testHealthCheckFuncWrapper func(ctx context.Context, newStream func() (interface{}, error), update func(bool), service string) error
+	extraDialOption            []grpc.DialOption
+}
+
+func setupClient(c *clientConfig) (cc *grpc.ClientConn, r *manual.Resolver, deferFunc func(), err error) {
+	r, rcleanup := manual.GenerateAndRegisterManualResolver()
+	defer rcleanup()
+	var opts []grpc.DialOption
+	opts = append(opts, grpc.WithInsecure(), grpc.WithBalancerName(c.balancerName))
+	if c.testHealthCheckFuncWrapper != nil {
+		opts = append(opts, internal.WithHealthCheckFunc.(func(interface{}) grpc.DialOption)(c.testHealthCheckFuncWrapper))
+	}
+	for _, dopt := range c.extraDialOption {
+		opts = append(opts, dopt)
+	}
+	cc, err = grpc.Dial(r.Scheme()+":///test.server", opts...)
+	if err != nil {
+		return nil, nil, func() { rcleanup() }, fmt.Errorf("dial failed due to err: %v", err)
```

Comment:
For error cases you should not return a cleanup function - the caller should see an error and assume nothing needs to be done to clean up.